src/Microsoft.AspNet.Cors.Core/CorsEngine.cs

```diff
@@ -0,0 +1,187 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNet.Http;
+using Microsoft.Framework.Internal;
+using Microsoft.Framework.OptionsModel;
+
+namespace Microsoft.AspNet.Cors.Core
+{
+    /// <summary>
+    /// Default implementation of <see cref="ICorsEngine"/>.
+    /// </summary>
+    public class CorsEngine : ICorsEngine
+    {
+        private readonly CorsOptions _options;
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="CorsEngine"/>.
+        /// </summary>
+        /// <param name="options">The option model representing <see cref="CorsOptions"/>.</param>
+        public CorsEngine([NotNull] IOptions<CorsOptions> options)
+        {
+            _options = options.Options;
+        }
+
+        /// <summary>
+        /// Looks up a policy using the <paramref name="policyName"/> and then evaluates the policy using the passed in
+        /// <paramref name="context"/>.
+        /// </summary>
+        /// <param name="requestContext"></param>
+        /// <param name="policyName"></param>
+        /// <returns>A <see cref="CorsResult"/> which contains the result of policy evaluation and can be
+        /// used by the caller to set apporpriate response headers.</returns>
+        public CorsResult EvaluatePolicy([NotNull] HttpContext context, string policyName)
+        {
+            var policy = _options.GetPolicy(policyName);
+            return EvaluatePolicy(context, policy);
+        }
+
+        /// <inheritdoc />
+        public CorsResult EvaluatePolicy([NotNull] HttpContext context, [NotNull]CorsPolicy policy)
+        {
```

Comment:
This should be changed to call two different methods, `EvaluateRequest(context, policy, result)` and `EvaluatePreflightRequest(context, policy, result)`. The `EvaluateRequest` should follow the steps defined by http://www.w3.org/TR/cors/#resource-requests and `EvaluatePreflightRequest` should follow http://www.w3.org/TR/cors/#resource-preflight-requests.

EvaluatePolicy calls EvaluatePreflightRequest when http method is `OPTIONS` and the request headers contain `Access-Control-Request-Method`, otherwise EvaluateRequest.